fda/magnitude_shape_plot.py

```diff
@@ -0,0 +1,272 @@
+
+    if dim_weights is None:
+        dim_weights = np.ones(fdatagrid.ndim_image) / fdatagrid.ndim_image
+
+    if pointwise_weights is None:
+        pointwise_weights = np.ones(fdatagrid.ncol) / fdatagrid.ncol
+
+    # Calculation of the depth of each multivariate sample with the corresponding weight.
+    weighted_depth = depth * dim_weights
+    sample_depth = weighted_depth.sum(axis=-1)
+
+    # Obtaining the median sample Z, to caculate v(t) = {X(t) − Z(t)}/∥ X(t) − Z(t)∥
+    median_index = np.argmax(sample_depth)
+    median = fdatagrid.data_matrix[median_index]
+    v = fdatagrid.data_matrix - median
+    v_norm = la.norm(v, axis=-1, keepdims=True)
+    #To avoid ZeroDivisionError, the zeros are substituted by ones.
+    v_norm[np.where(v_norm == 0)] = 1
+    v_unitary = v / v_norm
+
+    # Calculation of the depth of each point of each sample with the corresponding weight.
+    weighted_depth_pointwise = depth_pointwise * dim_weights
+    sample_depth_pointwise = weighted_depth_pointwise.sum(axis=-1, keepdims=True)
+
+    # Calcuation directinal outlyingness
+    dir_outlyingness = (1 / sample_depth_pointwise - 1) * v_unitary
+
+    # Calcuation mean directinal outlyingness
+    pointwise_weights_1 = np.tile(pointwise_weights, (fdatagrid.ndim_image, 1)).T
+    weighted_dir_outlyingness = dir_outlyingness * pointwise_weights_1
+    mean_dir_outl = weighted_dir_outlyingness.sum(axis=1)
+
+    # Calcuation variation directinal outlyingness
+    mean_dir_outl_pointwise = np.repeat(mean_dir_outl, fdatagrid.ncol, axis=0).reshape(fdatagrid.shape)
+    norm = np.square(la.norm(dir_outlyingness - mean_dir_outl_pointwise, axis=-1))
+    weighted_norm = norm * pointwise_weights
+    variation_dir_outl = weighted_norm.sum(axis=1)
+
+    return mean_dir_outl, variation_dir_outl
+
+def magnitude_shape_plot(fdatagrid, ax=None, depth_method=modified_band_depth, dim_weights=None,
+                         pointwise_weights=None, alpha = 0.993, colormap=plt.cm.get_cmap('seismic'),
+                         color = 0.2 , outliercol = 0.8,
+                         xlabel = 'MO', ylabel='VO', title='MS-Plot'):
+
+    r"""Implementation of the magnitude-shape plot
+
+    This plot which is based on the calculation of the
+    :func:`directional outlyingness <fda.magnitude_shape_plot._directional_outlyingness>`
+    of each of the samples and serves as a visualization tool for the centrality
+    of curves. Furthermore, an outlier detection procedure is included.
+
+    The norm of the mean of the directional outlyingness (:math:`\lVert\mathbf{MO}\rVert`)
+    is plotted in the x-axis, and the variation of the directional outlyingness (:math:`VO`)
+    in the y-axis.
+
+    Considering :math:`\mathbf{Y} = \left(\mathbf{MO}^T, VO\right)^T`, the outlier detection method
+    is implemented as described below.
+
+    First, the square robust Mahalanobis distance of :math:`\mathbf{Y}` is calculated with the minimum
```

Comment:
I do not fully understand the algorithm, as written here.

Reply:
Updated.